src/set/api.rs

```diff
@@ -0,0 +1,706 @@
+	/// assert_eq!(set.last(), Some(2));
+	/// ```
+	#[inline]
+	pub fn last(&self) -> Option<usize> {
+		self.inner.last_one()
+	}
+
+	/// Removes the first element from the set and returns it, if any.
+	/// The first element is always the minimum element in the set.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::pop_first`](alloc::collections::BTreeSet::pop_first)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	/// set.insert(1);
+	/// while let Some(n) = set.pop_first() {
+	///     assert_eq!(n, 1);
+	/// }
+	/// assert!(set.is_empty());
+	/// ```
+	#[inline]
+	pub fn pop_first(&mut self) -> Option<usize> {
+		let first = self.first()?;
+		self.remove(first);
+		Some(first)
+	}
+
+	/// Removes the last element from the set and returns it, if any.
+	/// The last element is always the maximum element in the set.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::pop_last`](alloc::collections::BTreeSet::pop_last)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	/// set.insert(1);
+	/// while let Some(n) = set.pop_last() {
+	///     assert_eq!(n, 1);
+	/// }
+	/// assert!(set.is_empty());
+	/// ```
+	#[inline]
+	pub fn pop_last(&mut self) -> Option<usize> {
+		let last = self.last()?;
+		self.remove(last);
+		Some(last)
+	}
+
+	/// Adds a value to the set.
+	///
+	/// Returns whether the value was newly inserted. That is:
+	///
+	/// * If the set did not previously contain the value, `true` is returned.
+	/// * If the set already contained the value, `false` is returned.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::insert`](alloc::collections::BTreeSet::insert)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	///
+	/// assert!(set.insert(2));
+	/// assert!(!set.insert(2));
+	/// assert_eq!(set.len(), 1);
+	/// ```
+	#[inline]
+	pub fn insert(&mut self, value: usize) -> bool {
+		let old_len = self.inner.len();
+		if value >= old_len {
+			self.inner.resize(value + 1, false);
+			self.inner.set(value, true);
+			false
+		}
+		else {
+			!self.inner.replace(value, true)
+		}
+	}
+
+	/// If the set contains the value, removes it from the set.
+	/// Returns whether such an element was present.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::remove`](alloc::collections::BTreeSet::remove)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	///
+	/// set.insert(2);
+	/// assert!(set.remove(2));
+	/// assert!(!set.remove(2));
+	/// ```
+	#[inline]
+	pub fn remove(&mut self, value: usize) -> bool {
+		let old_len = self.inner.len();
+		if value >= old_len {
+			false
+		}
+		else {
+			let ret = self.inner.replace(value, false);
+
+			// NOTE: it's unclear how this affects performance and if we should
+			// do this automatically, or require it manually only
+			self.shrink_inner();
```

Comment:
I haven't put much thought into this at all, so, would love input on whether this is a good idea.